packages/web/app/src/lib/preflight/hooks/usePreflight.tsx

```diff
@@ -0,0 +1,323 @@
+      Kit.tryOr(
+        () => JSON.parse(latestEnvironmentVariablesRef.current),
+        // todo: find a better solution than blowing away the user's
+        // invalid localStorage state.
+        //
+        // For example if the user has:
+        //
+        // { "foo": "bar }
+        //
+        // Then when they "Run Script" it will be replaced with:
+        //
+        // {}
+        //
+        () => ({}),
+      );
+    const result: PreflightResultData = {
+      request: {
+        headers: [],
+      },
+      environmentVariables: resultEnvironmentVariablesDecoded,
+    };
+
+    if (isPreview === false && !isEnabled) {
+      return result;
+    }
+
+    const id = crypto.randomUUID();
+    setState(PreflightWorkerState.running);
+    const now = Date.now();
+    setLogs(prev => [
+      ...prev,
+      {
+        level: 'log',
+        message: 'Running script...',
+      },
+    ]);
+
+    try {
+      const contentWindow = iframeRef.current?.contentWindow;
+
+      if (!contentWindow) {
+        throw new Error('Could not load iframe embed.');
+      }
+
+      contentWindow.postMessage(
+        {
+          type: IFrameEvents.Incoming.Event.run,
+          id,
+          script,
+          // Preflight has read/write relationship with environment variables.
+          environmentVariables: result.environmentVariables,
+        } satisfies IFrameEvents.Incoming.EventData,
+        '*',
+      );
+
+      let isFinished = false;
+      const isFinishedD = Promise.withResolvers<void>();
+      const openedPromptIds = new Set<number>();
+
+      // eslint-disable-next-line no-inner-declarations
+      function setFinished() {
+        isFinished = true;
+        isFinishedD.resolve();
+      }
+
+      // eslint-disable-next-line no-inner-declarations
+      function closedOpenedPrompts() {
+        if (openedPromptIds.size) {
+          for (const promptId of openedPromptIds) {
+            prompt.closePrompt(promptId, null);
+          }
+        }
+      }
+
+      // eslint-disable-next-line no-inner-declarations
+      async function eventHandler(ev: IFrameEvents.Outgoing.MessageEvent) {
+        if (ev.data.type === IFrameEvents.Outgoing.Event.prompt) {
+          const promptId = ev.data.promptId;
+          openedPromptIds.add(promptId);
+          await prompt
+            .openPrompt({
+              id: promptId,
+              title: ev.data.message,
+              defaultValue: ev.data.defaultValue,
+            })
+            .then(value => {
+              if (isFinished) {
+                // ignore prompt response if the script has already finished
+                return;
+              }
+
+              openedPromptIds.delete(promptId);
+              contentWindow?.postMessage(
+                {
+                  type: IFrameEvents.Incoming.Event.promptResponse,
+                  id,
+                  promptId,
+                  value,
+                } satisfies IFrameEvents.Incoming.EventData,
```

Comment:
This is cool. I need to read up more on the newer typescript features
https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-9.html#the-satisfies-operator

Reply:
`satisfies` was a nice ergonomic addition because it finally allowed the ability to have type-safe `const` types (whereas previously one had to choose between those two things). Not revolutionary but one of the many paper cut improvements over the past year or two.

While various details of TS have improved I think you might not be as far behind as you might feel :) in that the core type system has not significantly changed since perhaps the introduction of template literal types.

It feels to me like TS has significantly improved in meta ways, for example it made milestone improvements on ESM support.